common/lib/xmodule/xmodule/js/src/video/09_completion.js

```diff
@@ -0,0 +1,123 @@
+(function(define) {
+    'use strict';
+    define('video/09_completion.js', [], function() {
+    /**
+     * Completion handler
+     * @exports video/09_completion.js
+     * @constructor
+     * @param {Object} state The object containing the state of the video
+     * @return {jquery Promise}
+     */
+        var VideoCompletionHandler = function(state) {
+            if (!(this instanceof VideoCompletionHandler)) {
+                return new VideoCompletionHandler(state);
+            }
+            this.state = state;
+            this.state.completionHandler = this;
+            this.initialize();
+            return $.Deferred().resolve().promise();
+        };
+
+        VideoCompletionHandler.prototype = {
+            destroy: function() {
+                this.el.remove();
+                this.state.el.off('destroy', this.destroy);
+                delete this.state.completionHandler;
+            },
```

Comment:
This is all a bit mysterious to me, but I can see there's [other JS libraries doing something similar|https://github.com/open-craft/edx-platform/blob/1b9342faaa0ca8c0c6f118e70cc56c0ee10f2699/common/lib/xmodule/xmodule/js/src/video/09_events_plugin.js#L34-L37], so 👍

Reply:
Yeah.  I admit this part was copypasta from one of the other scripts.  I think it's just making sure that when the player gets destroyed (maybe when navigating away from the vertical?) all the resources get cleaned up appropriately--no dangling references causing memory leaks.  But I didn't track it all down.